[inductor] autotune benchmark support for cpu

jgong5 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124021
* #125736
* __->__ #125159

This PR adds the autotune Infrastructure for CPU. It generalizes and extends `BenchmarkRequest` with CPU support and C++ module loader. A `do_bench_cpu` util function is added for benchmarking functions on CPU with warmups and returns the median number from multiple trials.

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @ezyang @msaroufim @bdhirsh @anijain2305 @chauhang @voznesenskym @penguinwu @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

Commit:
Update

[ghstack-poisoned]

Commit:
Update

[ghstack-poisoned]

Commit:
Update

[ghstack-poisoned]

Commit:
Update

[ghstack-poisoned]

Commit:
Update

[ghstack-poisoned]

Commit:
Update

[ghstack-poisoned]

Commit:
Update

[ghstack-poisoned]

Commit:
Update

[ghstack-poisoned]

Commit:
Update on "[inductor] autotune benchmark support for cpu"




cc ezyang msaroufim bdhirsh anijain2305 chauhang voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire

[ghstack-poisoned]

jansel commented on third_party/ideep:
Do submodule update in a different PR.

jgong5 replied:
Thanks. A wrong "git add". Fixed.

shunting314 commented on torch/_inductor/ir.py:
```diff
@@ -3627,7 +3628,10 @@ def __init__(self, name, input_nodes, layout):
+        if is_cpu_device(args):
+            return do_bench_cpu(lambda: algo(*args, out=out))
+        else:
+            return do_bench(lambda: algo(*args, out=out))
```
Can we have this device dispatch logic inside do_bench?

Basically, rename current do_bench to do_bench_gpu. And create a new do_bench function that dispatches to do_bench_cpu/do_bench_gpu based on devices.

jgong5 replied:
Thanks. I did the refactoring in a separate PR: https://github.com/pytorch/pytorch/pull/125736

shunting314 commented on torch/_inductor/autotune_process.py:
```diff
@@ -763,6 +791,71 @@ def __str__(self) -> str:
+        return do_bench_cpu(fn)
+
+
+class CppBenchmarkRequest(CPUDeviceBenchmarkRequest):
```
hmm, where is this being used?

Commit:
Update on "[inductor] autotune benchmark support for cpu"


This PR adds the autotune Infrastructure for CPU. It generalizes and extends `BenchmarkRequest` with CPU support and C++ module loader. A `do_bench_cpu` util function is added for benchmarking functions on CPU with warmups and returns the median number from multiple trials.

cc gujinghui PenghuiCheng XiaobingSuper jianyuh mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen snadampal ezyang msaroufim bdhirsh anijain2305 chauhang voznesenskym penguinwu EikanWang zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire

[ghstack-poisoned]

Commit:
Update on "[inductor] autotune benchmark support for cpu"


This PR adds the autotune Infrastructure for CPU. It generalizes and extends `BenchmarkRequest` with CPU support and C++ module loader. A `do_bench_cpu` util function is added for benchmarking functions on CPU with warmups and returns the median number from multiple trials.

cc gujinghui PenghuiCheng XiaobingSuper jianyuh mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen snadampal ezyang msaroufim bdhirsh anijain2305 chauhang voznesenskym penguinwu EikanWang zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire

[ghstack-poisoned]

Commit:
Update on "[inductor] autotune benchmark support for cpu"


This PR adds the autotune Infrastructure for CPU. It generalizes and extends `BenchmarkRequest` with CPU support and C++ module loader. A `do_bench_cpu` util function is added for benchmarking functions on CPU with warmups and returns the median number from multiple trials.

cc gujinghui PenghuiCheng XiaobingSuper jianyuh mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen snadampal ezyang msaroufim bdhirsh anijain2305 chauhang voznesenskym penguinwu EikanWang zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire

[ghstack-poisoned]

jgong5 commented:
@pytorchbot merge

jerryzh168 commented:
we do have api depend on `do_bench` btw: https://github.com/pytorch/ao/blob/main/torchao/quantization/autoquant.py#L200 is this not a public API?

jgong5 commented:
> we do have api depend on `do_bench` btw: https://github.com/pytorch/ao/blob/main/torchao/quantization/autoquant.py#L200 is this not a public API?

I thought it was only used by inductor internal. You may have to rename it to `do_bench_gpu` now since the usage is cuda specific?
